sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -184,7 +183,9 @@ class Dataset[T] private[sql](
       // to happen right away to let these side effects take place eagerly.
       case p if hasSideEffects(p) =>
         LogicalRDD(queryExecution.analyzed.output, queryExecution.toRdd)(sparkSession)
-      case Union(children) if children.forall(hasSideEffects) =>
+      case Union(children) if children.exists(hasSideEffects) =>
```

Comment:
We catch `Union` here is not for queries like `INSERT ... UNION INSERT ...`, but for the multi insert syntax, see https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-InsertingdataintoHiveTablesfromqueries

So we don't need to handle `UNION ALL` here.